elastic/kibana
# [Enterprise Search] Add read-only mode interceptor and error handler

## Summary

This PR:

- Client-side:
  - Adds a layout-level read-only mode callout (that updates dynamically on every page visit/action)
  - Initializes the app with `readOnlyMode` from our `/api/ent/v1/internal/client_config` endpoint, if that's passed as true to us
  - Adds a [http interceptor](https://github.com/elastic/kibana/blob/master/docs/development/core/public/kibana-plugin-core-public.httpsetup.intercept.md) that listens to every call (specifically calls to our Enterprise Search APIs), and disables or enables read-only mode state if an `x-ent-search-read-only-mode` header is present and set to `true`
- Server-side
  - Sets response headers for our Kibana API endpoints (currently just the `x-ent-search-read-only-mode` header)
  - Updates all responses to pass back these headers (this allows the client-side interceptor behavior to work)
  - To enable error flash messages for POST/PUT/DELETE actions: Adds a special catch/check for the `x-ent-search-read-only-mode` header and adds a new handler for read only mode errors (since our current handler gobbles all 5xx server requests for security)

There are a good number of moving parts here, so I recommend following along by commit if possible! 🙏  closes https://github.com/elastic/app-search-team/issues/1193

### Screencaps

#### Standalone UI (for context):

<img width="1165" alt="" src="https://user-images.githubusercontent.com/549407/93284074-b3a78500-f786-11ea-9833-6bbfb8ac1ae9.png">

#### Kibana UI:

<img width="1366" alt="" src="https://user-images.githubusercontent.com/549407/93285542-f4ed6400-f789-11ea-8461-44ba6c176fb0.png">

## QA

- Setup
  - Have Enterprise Search running locally in a separate terminal
  - Start RO mode: `curl -X PUT 'http://localhost:3002/api/ent/v1/internal/read_only_mode' -d '{ "enabled": true }' -H "Content-Type: application/json" --user elastic:changeme`
- Testing RO mode on load:
  - Navigate to App Search
  - [x] Confirm the warning "Enterprise Search is in read-only mode" callout is present
- Testing dynamic updates to RO mode:
  - Stop RO mode: `curl -X PUT 'http://localhost:3002/api/ent/v1/internal/read_only_mode' -d '{ "enabled": false }' -H "Content-Type: application/json" --user elastic:changeme`
  - Navigate to Enterprise Search via breadcrumb
  - Navigate to App Search via product card
  - [x] Confirm that after the engines are done loading, the warning "Enterprise Search is in read-only mode" callout disappears 

### Checklist
